### ISSUE TITLE
Docker 1.6.2 pull race condition eats dependent layers

### ISSUE THREAD

**Issue by aaron42net** [coreos/bugs#377]
After an auto-update on "stable" from 668.3.0 to 681.0.0 in EC2, most of my images are broken.  Symptom:

```
 $ docker run -t -i 6eb041742985 /bin/bash
exec: "/bin/bash": stat /bin/bash: no such file or directory
FATA[0000] Error response from daemon: Cannot start container 199b0e74117dca1660bd17c9ba497697d398069ace47cd805bfe25a97159dbe7: [8] System error: exec: "/bin/bash": stat /bin/bash: no such file or directory
```

Looking through an export of this container, it looks like some (but not all) files are missing from our private base Debian layer, including /bin/bash.

I terminated this instance, and auto-scaling launched a new instance with 633.1.0.  After fleetctl starting my app and waiting for it to boot (which worked normally), I ran "sudo update_engine_client -update" and "sudo reboot", with the same result of corrupted images.

I do format ephemeral storage as btrfs with the following, which might be related:

```
coreos:
  units:
    - name: format-ephemeral.service
      command: start
      content: |
        [Unit]
        Description=Formats the ephemeral drive
        [Service]
        Type=oneshot
        RemainAfterExit=yes
        ExecStart=/usr/sbin/wipefs -f /dev/xvdb
        ExecStart=/usr/sbin/mkfs.btrfs -f /dev/xvdb
    - name: var-lib-docker.mount
      command: start
      content: |
        [Unit]
        Description=Mount ephemeral to /var/lib/docker
        Requires=format-ephemeral.service
        After=format-ephemeral.service
        Before=docker.service
        [Mount]
        What=/dev/xvdb
        Where=/var/lib/docker
        Type=btrfs
```

What additional detail can I provide?

**Comment by crawford** [comment]
@aaron42net is there any correlation between the missing files and which layer introduced them? I see that Docker was upgraded from 1.5.0 to 1.6.2. That might have something to do with this.

**Comment by crawford** [comment]
@aaron42net can you also provide the options with which you are invoking Docker?

**Comment by marineam** [comment]
@aaron42net also, could you post the journal from a failed boot? Want to make sure that there isn't something else going on like docker starting before var-lib-docker.mount that is confusing things.

**Comment by aaron42net** [comment]
Here's "journalctl -b" up until the point where the first app fails with that error: https://www.die.net/tmp/3d7e76d1a869ccec.log/failed-boot.log

And here's the service section of the fleet unit for the first one that failed, registrator.service:

```
[Service]
User=core
EnvironmentFile=/etc/environment
EnvironmentFile=/etc/environment.good
ExecStartPre=-/usr/bin/docker rm %n
ExecStartPre=-/usr/bin/docker pull goodinc/coreos:registrator
ExecStart=/usr/bin/docker run --name %n -m 1g -h ${HOSTNAME} -v /var/run/docker.sock:/tmp/docker.sock -e GOOD_ENV=${GOOD_ENV} -e ETCD=${ETCD} -e ETCD_TTL=${ETCD_TTL} -e COREOS_PRIVATE_IPV4=${COREOS_PRIVATE_IPV4} goodinc/coreos:registrator
ExecStop=/usr/bin/docker stop %n
TimeoutStartSec=5m
Restart=always
StartLimitInterval=5m
```

goodinc/coreos:registrator is layered atop goodinc/coreos:go, which is layered atop goodinc/coreos:debian, which in turn is layered atop debian:jessie (from 4 weeks ago).  Only 10532 files of 16730 are present in the broken goodinc/coreos:debian image.  Looking at the list of files missing, it looks like everything I install in goodinc/coreos:debian is there, but maybe nothing from debian:jessie.

**Comment by crawford** [comment]
```
Jun 11 12:47:27 ip-10-123-223-47.ec2.internal docker[854]: Pulling repository goodinc/coreos
Jun 11 12:47:31 ip-10-123-223-47.ec2.internal docker[854]: 2434ed458ced: Pulling image (registrator) from goodinc/coreos
Jun 11 12:47:31 ip-10-123-223-47.ec2.internal docker[854]: 2434ed458ced: Pulling image (registrator) from goodinc/coreos, endpoint: https://registry-1.docker.io/v1/
Jun 11 12:47:34 ip-10-123-223-47.ec2.internal docker[854]: 2434ed458ced: Pulling dependent layers
Jun 11 12:47:46 ip-10-123-223-47.ec2.internal docker[854]: 3cb35ae859e7: Pulling metadata
Jun 11 12:47:48 ip-10-123-223-47.ec2.internal docker[854]: 3cb35ae859e7: Pulling fs layer
Jun 11 12:47:51 ip-10-123-223-47.ec2.internal docker[854]: 3cb35ae859e7: Error downloading dependent layers
Jun 11 12:47:51 ip-10-123-223-47.ec2.internal docker[854]: 2434ed458ced: Error pulling image (registrator) from goodinc/coreos, endpoint: https://registry-1.docker.io/v1/, ApplyLayer exit status 1 stdout:  stderr: lchown /bin/dash: no such file or directory
Jun 11 12:47:51 ip-10-123-223-47.ec2.internal docker[854]: 2434ed458ced: Error pulling image (registrator) from goodinc/coreos, ApplyLayer exit status 1 stdout:  stderr: lchown /bin/dash: no such file or directory
Jun 11 12:47:51 ip-10-123-223-47.ec2.internal docker[854]: time="2015-06-11T12:47:51Z" level=fatal msg="Error pulling image (registrator) from goodinc/coreos, ApplyLayer exit status 1 stdout:  stderr: lchown /bin/dash: no such file or directory"
```

Here is the source of the error in the logs. It's not clear _exactly_ what's failing, but it doesn't look like it's specific to running on ephemeral storage.

**Comment by crawford** [comment]
@aaron42net can you test if you are able to pull this image from Docker 1.6.2 not on CoreOS? And then, can you test if you are able to pull that image with an older version of CoreOS (try 668.3.0 with Docker 1.5.0)?

**Comment by aaron42net** [comment]
See above.  These images do work fine with 668.3.0 and 633.1.0 on EC2.

They also work correctly on CoreOS 695.0.0 and 681.0.0 under VirtualBox on my Mac, but using the default ext4+overlayfs instead of btrfs.

I don't have any experience with running Docker outside CoreOS.

**Comment by aaron42net** [comment]
I tried using the 681.0.0 AMI in EC2 directly, rather than upgrading to it. I also switched to ext4+overlayfs instead of btrfs.  Same result.

**Comment by crawford** [comment]
That makes me think this is solely a Docker issue. Would it be possible for you to give me a copy of your image, or a subset of the layers? I'd like to try to reproduce the issue myself.

**Comment by aaron42net** [comment]
I ran docker build/push on all of our images from 681.0.0 on my Mac, and now I can no longer reproduce this issue in EC2.  My current theory is that whichever docker.io image cache I was hitting from EC2 us-east-1 had a corrupted debian:jessie layer, but I don't really have a way to validate that, nor understand why docker 1.5.0 was different than 1.6.2.

This unfortunately also means that I no longer have the broken image, sadly.

Thanks for your help.  I'm closing for now.

**Comment by crawford** [comment]
Very interesting. I'm glad you got it sorted out.

I wonder if there was a strange interaction between the version of Docker which originally created the image and 1.6.2 (which actually ran it). Maybe rebuilding the image with a newer version of Docker fixed the issue. Oh well.

**Comment by aaron42net** [comment]
So sadly, this issue has returned for me.  And I'm now wondering if this is a race-condition or bug with docker pull.

On startup, I have 20 fleet units trying to pull images and then run containers from them simultaneously, all of which are layered atop the debian:jessie layer.  If they fail, they rm their containers (but not the images) and try again.

While that was going on at startup, I did this sequence, which suggests that the image was fine, and then got downloaded again, then was broken:

```
core@ip-10-157-118-54 ~ $ docker pull goodinc/coreos:debian
debian: Pulling from goodinc/coreos
61b3964dfa70: Already exists 
f5224fc54ad2: Already exists 
713c1d9ccb27: Already exists 
432092d2ba57: Already exists 
Digest: sha256:4e4735b82d35d9ac855f737c28161ab5f5fa4dad71a304adcce24a5150665c9a
Status: Downloaded newer image for goodinc/coreos:debian

core@ip-10-157-118-54 ~ $ docker pull goodinc/coreos:debian
debian: Pulling from goodinc/coreos
61b3964dfa70: Already exists 
f5224fc54ad2: Already exists 
713c1d9ccb27: Already exists 
432092d2ba57: Already exists 
Digest: sha256:4e4735b82d35d9ac855f737c28161ab5f5fa4dad71a304adcce24a5150665c9a
Status: Image is up to date for goodinc/coreos:debian

core@ip-10-157-118-54 ~ $ docker pull goodinc/coreos:debian
debian: Pulling from goodinc/coreos
61b3964dfa70: Already exists 
f5224fc54ad2: Already exists 
713c1d9ccb27: Already exists 
432092d2ba57: Already exists 
Digest: sha256:4e4735b82d35d9ac855f737c28161ab5f5fa4dad71a304adcce24a5150665c9a
Status: Image is up to date for goodinc/coreos:debian

core@ip-10-157-118-54 ~ $ docker run -t -i goodinc/coreos:debian /bin/bash
Unable to find image 'goodinc/coreos:debian' locally
ps auxwwdebian: Pulling from goodinc/coreos
61b3964dfa70: Extracting [=>                                                 ] 1.049 MB/51.36 MB
61b3964dfa70: Error downloading dependent layers 
432092d2ba57: Error pulling image (debian) from goodinc/coreos, ApplyLayer exit status 1 stdout:  stderr: lchown /bin/bash: no such file or directory 
FATA[0023] Error pulling image (debian) from goodinc/coreos, ApplyLayer exit status 1 stdout:  stderr: lchown /bin/bash: no such file or directory 

core@ip-10-157-118-54 ~ $ docker run -t -i goodinc/coreos:debian /bin/bash
Unable to find image 'goodinc/coreos:debian' locally
debian: Pulling from goodinc/coreos
61b3964dfa70: Already exists 
f5224fc54ad2: Already exists 
713c1d9ccb27: Already exists 
432092d2ba57: Already exists 
Digest: sha256:4e4735b82d35d9ac855f737c28161ab5f5fa4dad71a304adcce24a5150665c9a
Status: Image is up to date for goodinc/coreos:debian
FATA[0009] Error response from daemon: No such image: goodinc/coreos:debian (tag: debian) 
```

If you feel like experimenting, I've granted "crawford" on docker hub read-only access to goodinc/coreos, which hopefully is you.

**Comment by crawford** [comment]
@aaron42net I am abcrawf on docker hub! I don't know who that impostor is.

**Comment by aaron42net** [comment]
Aha. I swapped out crawford for abcrawf on docker hub.

**Comment by aaron42net** [comment]
Okay, I've narrowed this down.  See https://github.com/aaron42net/dockertest for my simplified test case, and instructions to reproduce.

Multiple simultaneous docker pulls that depend on the same underlying layers will corrupt them.  The only safe way to do docker pulls with docker 1.6.2 is sequentially, which is hard to guarantee with fleet starting multiple units.

**Comment by aaron42net** [comment]
I've submitted https://github.com/docker/docker/issues/13971 since this appears to be a docker issue.  It's bad enough for me that I'm having trouble using CoreOS stable right now, though.

**Comment by scole-scea** [comment]
Try using `flock` maybe? In our own case, I `flock`-protect all docker pulls (with a read lock) because I've got dangling-image removal (which grabs a `flock` write lock) in ExecStartPost, and docker unreliably calculates the dangling list while pulls are going on.

If docker 1.6.2 has pull problems as well, we might switch those pulls to write locks as well.

**Comment by aaron42net** [comment]
Thanks for the suggestion. As a workaround, I've added the following to all of my fleet units, and it does serialize pulls and prevent the corruption.  It's also quite a bit slower to start up.

```
ExecStartPre=-/usr/bin/flock /home/core/pull.lock /usr/bin/timeout 5m /usr/bin/docker pull ...
```

**Comment by scole-scea** [comment]
Awesome. I'd put your lock in `/var/lock` rather than `/home/core`, but then again, workarounds are supposed to be hacks anyway. :smile:

**Comment by crawford** [comment]
@aaron42net has this been "resolved" now that jessie was [rebuilt with 1.6.0](https://github.com/docker/docker/issues/13971#issuecomment-113648832)?

**Comment by aaron42net** [comment]
@crawford I just rebuilt the aaron42net/dockertest images with the latest debian:jessie and I can still reproduce.

**Comment by scole-scea** [comment]
This just hit me, too (though it took surprisingly long to rear its head, given when the issue first appeared). Really feels like a docker bug, though, not a CoreOS bug.

**Comment by crawford** [comment]
We've shipped Docker 1.8 and 1.7 in all of our channels.
